src/circom/mod.rs

```diff
@@ -0,0 +1,135 @@
+    let witness = witness_calculator
+      .calculate_witness(input.clone(), true)
+      .expect("msg");
+
+    self.witness = Some(witness);
+  }
+}
+
+impl<F: PrimeField> Circuit<F> for SpartanCircuit<F> {
+  fn synthesize<CS: ConstraintSystem<F>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
+    let _ = circom_scotia::synthesize(cs, self.r1cs.clone(), self.witness).unwrap();
+
+    Ok(())
+  }
+}
+
+#[allow(dead_code)]
+pub fn setup<G: Group, S: RelaxedR1CSSNARKTrait<G>>(
+  circuit: SpartanCircuit<<G as Group>::Scalar>,
+) -> (ProverKey<G, S>, VerifierKey<G, S>) {
+  SNARK::<G, S, SpartanCircuit<<G as Group>::Scalar>>::setup(circuit).unwrap()
+}
+
+#[allow(dead_code)]
+pub fn generate_proof<G: Group, S: RelaxedR1CSSNARKTrait<G>>(
+  pk: ProverKey<G, S>,
+  circuit: &mut SpartanCircuit<<G as Group>::Scalar>,
+  input: Vec<(String, Vec<<G as Group>::Scalar>)>,
+  wtns_path: PathBuf,
+) -> Result<SNARK<G, S, SpartanCircuit<<G as traits::Group>::Scalar>>, SpartanError> {
+  circuit.compute_witness(input, wtns_path);
+  SNARK::prove(&pk, circuit.clone())
+}
+
+#[allow(dead_code)]
+pub fn create_snark<G: Group, S: RelaxedR1CSSNARKTrait<G>>(
+  r1cs_path: PathBuf,
+  wtns_path: PathBuf,
+  input: Vec<(String, Vec<<G as Group>::Scalar>)>,
+) -> (
+  ProverKey<G, S>,
+  VerifierKey<G, S>,
+  Result<SNARK<G, S, SpartanCircuit<<G as traits::Group>::Scalar>>, SpartanError>,
+) {
+  let mut circuit = SpartanCircuit::new(r1cs_path);
+  let (pk, vk) =
+    SNARK::<G, S, SpartanCircuit<<G as Group>::Scalar>>::setup(circuit.clone()).unwrap();
+  circuit.compute_witness(input, wtns_path);
+  let proof = SNARK::prove(&pk, circuit.clone());
+  (pk, vk, proof)
+}
+
+#[cfg(test)]
+mod test {
+  use super::{create_snark, generate_proof, setup, SpartanCircuit};
+  use crate::{provider::bn256_grumpkin::bn256, traits::Group};
+  use std::env::current_dir;
+
+  #[test]
+  fn test_spartan_snark() {
+    type G = bn256::Point;
+    type EE = crate::provider::ipa_pc::EvaluationEngine<G>;
+    type S = crate::spartan::snark::RelaxedR1CSSNARK<G, EE>;
+
+    let root = current_dir().unwrap().join("examples/cube");
+    let r1cs_path = root.join("cube.r1cs");
+    let wtns_path = root.join("cube.wasm");
+
+    let arg_x = ("x".into(), vec![<G as Group>::Scalar::from(2)]);
+    let arg_y = ("y".into(), vec![<G as Group>::Scalar::from(8)]);
+    let input = vec![arg_x, arg_y];
+
+    let (_, vk, res) = create_snark::<G, S>(r1cs_path, wtns_path, input);
+    assert!(res.is_ok());
+
+    let snark = res.unwrap();
+    assert!(snark.verify(&vk).is_ok());
+  }
+
+  #[test]
+  fn test_spartan_snark_fail() {
+    type G = bn256::Point;
+    type EE = crate::provider::ipa_pc::EvaluationEngine<G>;
+    type S = crate::spartan::snark::RelaxedR1CSSNARK<G, EE>;
+
+    let root = current_dir().unwrap().join("examples/cube");
+    let r1cs_path = root.join("cube.r1cs");
+    let wtns_path = root.join("cube.wasm");
+    let mut circuit = SpartanCircuit::new(r1cs_path);
```

Comment:
There's a simplification here. We can just pass `r1cs_path` to setup (which returns `(pk,vk)`) and the `setup` can internally create a `SpartanCircuit` object (this will hide the struct from the caller and reduces code that someone needs to write). Isn't it?